File: ios/CellularChatCore/Sources/CellularChatCore/SecureSession.swift
import Foundation

/// Session driver (PROTOCOL_V2.md §8): IKpsk2 with pinned-peer verification
/// BEFORE any response, then per-direction AEAD envelopes with seq==counter and
/// sid checks, version checks, and unknown-msgType handling.
public final class SecureSession {

    public enum Role { case initiator, responder }

    public enum Inbound: Equatable {
        case message(msgType: UInt64, body: CBOR)
        case ignored   // unknown msgType >= 128 (forward-compatible extension)
    }

    private let role: Role
    private let pinnedPeer: [UInt8]

    private var hs: HandshakeState?
    private var sendCipher: CipherState?
    private var recvCipher: CipherState?
    private var handshakeComplete = false

    private var sidValue: [UInt8]?
    private var didFirstSend = false
    private var didFirstReceive = false

    public var sid: [UInt8]? { sidValue }
    public private(set) var handshakeHash: [UInt8]?
    var sendKey: [UInt8]? { sendCipher?.k }   // internal: for vector assertions
    var recvKey: [UInt8]? { recvCipher?.k }

    /// - Parameters:
    ///   - localStaticPriv: this device's pair static private key.
    ///   - pinnedPeerStaticPub: the pinned peer static public key (§2).
    ///   - sid: required for the initiator; the responder adopts the initiator's.
    public init(role: Role,
                prologue: [UInt8],
                sessionPsk: [UInt8],
                localStaticPriv: [UInt8],
                pinnedPeerStaticPub: [UInt8],
                ephemeralPriv: [UInt8],
                sid: [UInt8]? = nil) throws {
        self.role = role
        self.pinnedPeer = pinnedPeerStaticPub
        self.sidValue = sid
        switch role {
        case .initiator:
            hs = try HandshakeState(
                pattern: .ikpsk2, initiator: true, prologue: prologue,
                s: localStaticPriv, e: ephemeralPriv, rs: pinnedPeerStaticPub, psk: sessionPsk)
        case .responder:
            hs = try HandshakeState(
                pattern: .ikpsk2, initiator: false, prologue: prologue,
                s: localStaticPriv, e: ephemeralPriv, rs: nil, psk: sessionPsk)
        }
    }

    // MARK: Handshake

    /// Initiator: produce message 1 record. Responder: produce message 2 record
    /// (only after `readHandshake` accepted message 1 and pinning passed).
    public func writeHandshake() throws -> [UInt8] {
        guard let hs, !handshakeComplete else { throw ProtocolError.wrongMessagePhase }
        let msg = try hs.writeMessage([])
        guard msg.count <= RecordLimits.maxNoiseMessage else { throw ProtocolError.oversizeRecord }
        let record = Records.make(.sessionHandshake, payload: msg)
        if role == .responder {
            finishHandshake()   // responder completes after writing msg2
        }
        return record
    }

    /// Responder: read message 1, then verify the transmitted static equals the
    /// pinned peer key BEFORE any response. Initiator: read message 2.
    public func readHandshake(_ record: [UInt8]) throws {
        guard let hs, !handshakeComplete else { throw ProtocolError.wrongMessagePhase }
        let (type, payload) = try Records.parse(record)
        guard type == .sessionHandshake else { throw ProtocolError.wrongMessagePhase }
        _ = try hs.readMessage(payload)
        if role == .responder {
            guard let learned = hs.rs,
                  Crypto.constantTimeEqual(learned, pinnedPeer) else {
                throw ProtocolError.pinnedKeyMismatch
            }
        } else {
            finishHandshake()   // initiator completes after reading msg2
        }
    }

    private func finishHandshake() {
        guard let hs else { return }
        handshakeHash = hs.handshakeHash
        let (c1, c2) = hs.split()
        switch role {
        case .initiator:
            sendCipher = c1   // initiator -> responder
            recvCipher = c2
        case .responder:
            recvCipher = c1
            sendCipher = c2   // responder -> initiator
        }
        handshakeComplete = true
    }

    // MARK: Transport

    public func send(msgType: UInt64, body: CBOR) throws -> [UInt8] {
        guard handshakeComplete, let sendCipher, let sid = sidValue else {
            throw ProtocolError.wrongMessagePhase
        }
        if !didFirstSend && msgType != SessionMsgType.sessionReady.rawValue {
            throw ProtocolError.schemaViolation   // first transport message MUST be session_ready
        }
        let env = SessionEnvelope(msgType: msgType, seq: sendCipher.n, sid: sid, body: body)
        let plain = env.encoded()
        guard plain.count <= RecordLimits.maxPlaintextEnvelope else { throw ProtocolError.oversizeRecord }
        let ct = try sendCipher.encryptWithAd([], plain)
        didFirstSend = true
        return Records.make(.sessionTransport, payload: ct)
    }

    public func receive(_ record: [UInt8]) throws -> Inbound {
        let (type, payload) = try Records.parse(record)
        guard type == .sessionTransport else {
            // Handshake record after completion, or any wrong-phase record (§14).
            throw ProtocolError.wrongMessagePhase
        }
        guard handshakeComplete, let recvCipher else { throw ProtocolError.wrongMessagePhase }

        let expectedCounter = recvCipher.n
        let plain = try recvCipher.decryptWithAd([], payload)   // throws aeadFailure / counterOverflow
        let env = try SessionEnvelope.decode(plain)

        guard env.seq == expectedCounter else { throw ProtocolError.counterMismatch }
        if let sid = sidValue {
            guard Crypto.constantTimeEqual(env.sid, sid) else { throw ProtocolError.sidMismatch }
        } else {
            sidValue = env.sid   // responder adopts the initiator's sid
        }

        if !didFirstReceive && env.msgType != SessionMsgType.sessionReady.rawValue {
            throw ProtocolError.schemaViolation
        }
        didFirstReceive = true

        // Version negotiation: any field claiming < 2 is rejected (§8).
        if env.msgType == SessionMsgType.sessionReady.rawValue {
            guard let maxVersion = env.body.value(forKey: 3)?.asUInt, maxVersion >= 2 else {
                throw ProtocolError.unsupportedVersion
            }
        }

        // Unknown-message rule (§8): < 128 fatal, >= 128 ignored.
        if SessionMsgType(rawValue: env.msgType) == nil {
            if env.msgType < 128 { throw ProtocolError.schemaViolation }
            return .ignored
        }
        return .message(msgType: env.msgType, body: env.body)
    }
}

File: android/app/src/main/java/com/cellularchat/app/core/protocol/SecureSession.kt
package com.cellularchat.app.core.protocol

import com.cellularchat.app.core.ProtocolException
import com.cellularchat.app.core.ReasonCodes
import com.cellularchat.app.core.cbor.CborInt
import com.cellularchat.app.core.cbor.CborMap
import com.cellularchat.app.core.crypto.Derivations
import com.cellularchat.app.core.noise.CipherState
import com.cellularchat.app.core.noise.HandshakeState
import com.cellularchat.app.core.noise.NoisePattern
import java.nio.charset.StandardCharsets
import java.security.MessageDigest

/**
 * Drives IKpsk2 (PROTOCOL_V2.md §8) for either role and then encrypts/decrypts
 * session transport envelopes with per-direction counters. Enforces the §14
 * rejection rules: pinned-static check before any response, seq==counter, sid
 * match, oversize, unknown-msgType policy, and the counter overflow guard.
 */
class SecureSession private constructor(
    private val role: Role,
    pairId: ByteArray,
    transportTag: String,
    private val localStaticPrivate: ByteArray,
    private val pinnedPeerStatic: ByteArray,
    pairRoot: ByteArray,
    private var sessionId: ByteArray?,
    fixedEphemeralPrivate: ByteArray?,
) {
    enum class Role { INITIATOR, RESPONDER }

    /** A received message; [ignored] is set for forward-compatible msgType ≥ 128. */
    data class Incoming(val envelope: SessionEnvelope, val ignored: Boolean)

    private val prologue = "cellfind/v2/session".toByteArray(StandardCharsets.US_ASCII) +
        pairId + transportTag.toByteArray(StandardCharsets.US_ASCII)

    private val handshake = HandshakeState(
        pattern = NoisePattern.IK_PSK2,
        initiator = role == Role.INITIATOR,
        prologue = prologue,
        psk = Derivations.sessionPsk(pairRoot),
        localStaticPrivate = localStaticPrivate,
        remoteStaticPublic = if (role == Role.INITIATOR) pinnedPeerStatic else null,
        fixedEphemeralPrivate = fixedEphemeralPrivate,
    )

    private var initToResp: CipherState? = null
    private var respToInit: CipherState? = null

    private val sendCipher: CipherState
        get() = (if (role == Role.INITIATOR) initToResp else respToInit)
            ?: throw ProtocolException("transport message before handshake completed")
    private val recvCipher: CipherState
        get() = (if (role == Role.INITIATOR) respToInit else initToResp)
            ?: throw ProtocolException("transport message before handshake completed")

    val isEstablished: Boolean get() = initToResp != null

    fun handshakeHash(): ByteArray = handshake.handshakeHash()

    // --- IKpsk2 handshake ---

    /** Initiator: produce record 0x02 (session_hello / msg1). */
    fun startHandshake(): ByteArray {
        require(role == Role.INITIATOR) { "only the initiator starts the handshake" }
        return Records.build(Records.SESSION_HANDSHAKE, handshake.writeMessage(ByteArray(0)))
    }

    /**
     * Responder: consume record 0x02 (msg1). Performs the pinning check on the
     * initiator's transmitted static BEFORE any response can be produced.
     */
    fun readHandshakeInit(record: ByteArray) {
        require(role == Role.RESPONDER) { "only the responder reads msg1" }
        expectRecordType(record, Records.SESSION_HANDSHAKE)
        handshake.readMessage(Records.payload(record))
        val transmitted = handshake.remoteStaticKey()
            ?: throw ProtocolException("initiator static key missing", ReasonCodes.IDENTITY_MISMATCH)
        if (!MessageDigest.isEqual(transmitted, pinnedPeerStatic)) {
            throw ProtocolException("initiator static does not match the pinned peer key", ReasonCodes.IDENTITY_MISMATCH)
        }
    }

    /** Responder: produce record 0x02 (session_auth / msg2) and complete. */
    fun writeHandshakeResponse(): ByteArray {
        require(role == Role.RESPONDER) { "only the responder writes msg2" }
        val record = Records.build(Records.SESSION_HANDSHAKE, handshake.writeMessage(ByteArray(0)))
        finishHandshake()
        return record
    }

    /** Initiator: consume record 0x02 (msg2) and complete. */
    fun readHandshakeResponse(record: ByteArray) {
        require(role == Role.INITIATOR) { "only the initiator reads msg2" }
        expectRecordType(record, Records.SESSION_HANDSHAKE)
        handshake.readMessage(Records.payload(record))
        finishHandshake()
    }

    private fun finishHandshake() {
        val (c1, c2) = handshake.splitStates ?: throw ProtocolException("handshake not complete")
        initToResp = c1
        respToInit = c2
    }

    /** (initiator→responder key, responder→initiator key) after the handshake. */
    fun transportKeys(): Pair<ByteArray, ByteArray> {
        val c1 = initToResp ?: throw ProtocolException("handshake not complete")
        val c2 = respToInit ?: throw ProtocolException("handshake not complete")
        return c1.keyBytes()!! to c2.keyBytes()!!
    }

    // --- session transport ---

    fun send(msgType: Long, body: CborMap): ByteArray {
        val cipher = sendCipher
        guardNonce(cipher)
        val sid = sessionId ?: throw ProtocolException("session id not established")
        val plaintext = SessionEnvelopeCodec.encode(SessionEnvelope(msgType, cipher.nonce, sid, body))
        if (plaintext.size > Records.MAX_PLAINTEXT_ENVELOPE) {
            throw ProtocolException("session envelope exceeds the plaintext limit")
        }
        val ciphertext = cipher.encryptWithAd(EMPTY_AD, plaintext)
        return Records.build(Records.SESSION_TRANSPORT, ciphertext)
    }

    /** Returns the decrypted envelope, or null when a msgType ≥ 128 is ignored. */
    fun receive(record: ByteArray): SessionEnvelope? {
        val incoming = receiveDetailed(record)
        return if (incoming.ignored) null else incoming.envelope
    }

    fun receiveDetailed(record: ByteArray): Incoming {
        expectRecordType(record, Records.SESSION_TRANSPORT)
        val cipher = recvCipher
        guardNonce(cipher)
        val expectedSeq = cipher.nonce
        val plaintext = cipher.decryptWithAd(EMPTY_AD, Records.payload(record))
        val envelope = SessionEnvelopeCodec.decode(plaintext)
        if (envelope.seq != expectedSeq) throw ProtocolException("seq does not match the counter")
        val currentSid = sessionId
        if (currentSid == null) {
            sessionId = envelope.sid
        } else if (!envelope.sid.contentEquals(currentSid)) {
            throw ProtocolException("session id mismatch")
        }
        if (envelope.msgType == SessionMsgType.SESSION_READY) {
            val maxVersion = (envelope.body[3L] as? CborInt)?.value
            if (maxVersion != null && maxVersion < 2) {
                throw ProtocolException("protocol downgrade below version 2", ReasonCodes.CAPABILITY_MISMATCH)
            }
        }
        if (envelope.msgType !in SessionMsgType.KNOWN) {
            if (envelope.msgType < 128) throw ProtocolException("unknown session msgType ${envelope.msgType}")
            return Incoming(envelope, ignored = true)
        }
        return Incoming(envelope, ignored = false)
    }

    private fun guardNonce(cipher: CipherState) {
        if (cipher.nonce >= MAX_NONCE) {
            throw ProtocolException("counter overflow; re-handshake required before 2^32")
        }
    }

    private fun expectRecordType(record: ByteArray, expected: Int) {
        if (Records.recordType(record) != expected) {
            throw ProtocolException("unexpected record type for this phase")
        }
    }

    /** Test hook to exercise the counter overflow guard without 2^32 sends. */
    internal fun debugSetSendCounter(value: Long) {
        sendCipher.setNonce(value)
    }

    companion object {
        private val EMPTY_AD = ByteArray(0)
        const val MAX_NONCE: Long = 1L shl 32

        fun initiator(
            pairId: ByteArray,
            transportTag: String,
            localStaticPrivate: ByteArray,
            pinnedPeerStatic: ByteArray,
            pairRoot: ByteArray,
            sid: ByteArray,
            fixedEphemeralPrivate: ByteArray? = null,
        ) = SecureSession(
            Role.INITIATOR, pairId, transportTag, localStaticPrivate, pinnedPeerStatic,
            pairRoot, sid, fixedEphemeralPrivate,
        )

        fun responder(
            pairId: ByteArray,
            transportTag: String,
            localStaticPrivate: ByteArray,
            pinnedPeerStatic: ByteArray,
            pairRoot: ByteArray,
            fixedEphemeralPrivate: ByteArray? = null,
        ) = SecureSession(
            Role.RESPONDER, pairId, transportTag, localStaticPrivate, pinnedPeerStatic,
            pairRoot, null, fixedEphemeralPrivate,
        )
    }
}
